File: packages/flutter_blue_plus_darwin/darwin/flutter_blue_plus_darwin/Sources/flutter_blue_plus_darwin/Proto.swift
// Copyright 2017-2023, Charles Weinberger & Paul DeMarco.
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import CoreBluetooth
import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif

// ─────────────────────────────────────────────────────────────────────────────
// Errors
// ─────────────────────────────────────────────────────────────────────────────

/// Wraps a CoreBluetooth NSError as a stable "cb_error" with the raw native
/// code as details.
func cbError(_ error: Error) -> PigeonError {
  let ns = error as NSError
  return PigeonError(code: "cb_error", message: ns.localizedDescription, details: Int64(ns.code))
}

func notConnectedError() -> PigeonError {
  return PigeonError(code: "not_connected", message: "device is disconnected", details: nil)
}

func deviceDisconnectedError() -> PigeonError {
  return PigeonError(code: "device_disconnected", message: "device is disconnected", details: nil)
}

func adapterOffError(_ state: CBManagerState) -> PigeonError {
  return PigeonError(
    code: "adapter_off",
    message: "bluetooth must be turned on. (\(cbManagerStateString(state)))",
    details: nil)
}

func unsupportedError(_ message: String) -> PigeonError {
  return PigeonError(code: "unsupported", message: message, details: nil)
}

func cbManagerStateString(_ state: CBManagerState) -> String {
  switch state {
  case .unknown: return "CBManagerStateUnknown"
  case .unsupported: return "CBManagerStateUnsupported"
  case .unauthorized: return "CBManagerStateUnauthorized"
  case .resetting: return "CBManagerStateResetting"
  case .poweredOn: return "CBManagerStatePoweredOn"
  case .poweredOff: return "CBManagerStatePoweredOff"
  @unknown default: return "unhandled"
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Enums
// ─────────────────────────────────────────────────────────────────────────────

func bmAdapterState(_ state: CBManagerState) -> BmAdapterStateEnum {
  switch state {
  case .unknown: return .unknown
  case .unsupported: return .unavailable
  case .unauthorized: return .unauthorized
  case .resetting: return .turningOn
  case .poweredOn: return .on
  case .poweredOff: return .off
  @unknown default: return .unknown
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// GATT tree
// ─────────────────────────────────────────────────────────────────────────────

func bmBluetoothDevice(_ peripheral: CBPeripheral) -> BmBluetoothDevice {
  return BmBluetoothDevice(
    address: peripheral.identifier.uuidString,
    platformName: peripheral.name)
}

func bmBluetoothService(_ service: CBService, in peripheral: CBPeripheral) -> BmBluetoothService {
  let characteristics = (service.characteristics ?? []).map { bmBluetoothCharacteristic($0) }

  // included (secondary) services are referenced by ref; their full trees are
  // emitted as separate top-level entries of the discovery result
  let includedServices = (service.includedServices ?? []).map { included in
    BmServiceRef(service: attributeId(included), parentService: attributeId(service))
  }

  return BmBluetoothService(
    id: attributeId(service),
    isPrimary: service.isPrimary,
    characteristics: characteristics,
    includedServices: includedServices)
}

func bmBluetoothCharacteristic(_ characteristic: CBCharacteristic) -> BmBluetoothCharacteristic {
  let descriptors = (characteristic.descriptors ?? []).map {
    BmBluetoothDescriptor(uuid: $0.uuid.uuidStr)
  }
  return BmBluetoothCharacteristic(
    id: attributeId(characteristic),
    descriptors: descriptors,
    properties: bmCharacteristicProperties(characteristic.properties))
}

func bmCharacteristicProperties(_ props: CBCharacteristicProperties) -> BmCharacteristicProperties {
  return BmCharacteristicProperties(
    broadcast: props.contains(.broadcast),
    read: props.contains(.read),
    writeWithoutResponse: props.contains(.writeWithoutResponse),
    write: props.contains(.write),
    notify: props.contains(.notify),
    indicate: props.contains(.indicate),
    authenticatedSignedWrites: props.contains(.authenticatedSignedWrites),
    extendedProperties: props.contains(.extendedProperties),
    notifyEncryptionRequired: props.contains(.notifyEncryptionRequired),
    indicateEncryptionRequired: props.contains(.indicateEncryptionRequired))
}

/// CBDescriptor values are typed (NSString / NSNumber / NSData) — convert to
/// raw bytes like the ObjC `descriptorToData`.
func descriptorValueData(_ descriptor: CBDescriptor) -> Data {
  guard let value = descriptor.value else { return Data() }
  if let string = value as? String {
    return string.data(using: .utf8) ?? Data()
  }
  if let number = value as? NSNumber {
    var intValue = Int32(truncatingIfNeeded: number.intValue)
    return withUnsafeBytes(of: &intValue) { Data($0) }
  }
  if let data = value as? Data {
    return data
  }
  return Data()
}

// ─────────────────────────────────────────────────────────────────────────────
// Scan advertisements
// ─────────────────────────────────────────────────────────────────────────────

func bmScanAdvertisement(
  address: String,
  peripheral: CBPeripheral?,
  advertisementData: [String: Any],
  rssi: NSNumber
) -> BmScanAdvertisement {
  let advName = advertisementData[CBAdvertisementDataLocalNameKey] as? String
  let connectable = advertisementData[CBAdvertisementDataIsConnectable] as? NSNumber
  let txPower = advertisementData[CBAdvertisementDataTxPowerLevelKey] as? NSNumber
  let manufData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data
  let serviceUuids = advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID]
  let serviceData = advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID: Data]

  // manufacturer data: first 2 bytes are the manufacturer id (little endian)
  var manufacturerData: [Int64: FlutterStandardTypedData] = [:]
  if let manufData = manufData, manufData.count >= 2 {
    let bytes = [UInt8](manufData.prefix(2))
    let manufId = UInt16(bytes[0]) | (UInt16(bytes[1]) << 8)
    let trimmed = manufData.subdata(in: 2..<manufData.count)
    manufacturerData[Int64(manufId)] = FlutterStandardTypedData(bytes: trimmed)
  }

  var serviceDataMap: [String: FlutterStandardTypedData] = [:]
  for (uuid, data) in serviceData ?? [:] {
    serviceDataMap[uuid.uuidStr] = FlutterStandardTypedData(bytes: data)
  }

  return BmScanAdvertisement(
    address: address,
    platformName: peripheral?.name,
    advName: advName,
    connectable: connectable?.boolValue ?? false,
    txPowerLevel: txPower.map { Int64(truncating: $0) },
    appearance: nil,  // not supported on iOS / macOS
    manufacturerData: manufacturerData,
    serviceData: serviceDataMap,
    serviceUuids: (serviceUuids ?? []).map { $0.uuidStr },
    rssi: Int64(truncating: rssi))
}

// ─────────────────────────────────────────────────────────────────────────────
// Scan filters (implemented by FBP, not by the OS)
// ─────────────────────────────────────────────────────────────────────────────

enum ScanFilters {
  static func hasCustomFilters(_ settings: BmScanSettings) -> Bool {
    return !settings.withRemoteIds.isEmpty
      || !settings.withNames.isEmpty
      || !settings.withKeywords.isEmpty
      || !settings.withMsd.isEmpty
      || !settings.withServiceData.isEmpty
  }

  static func isAnyFilterSet(_ settings: BmScanSettings) -> Bool {
    return !settings.withServices.isEmpty || hasCustomFilters(settings)
  }

  /// Filters are additive: an advertisement can match *any* filter.
  static func allows(
    _ settings: BmScanSettings,
    remoteId: String,
    advName: String?,
    advServices: [CBUUID]?,
    msd: Data?,
    serviceData: [CBUUID: Data]?
  ) -> Bool {
    if !isAnyFilterSet(settings) {
      return true
    }
    if foundService(settings.withServices, target: advServices) {
      return true
    }
    if foundRemoteId(settings.withRemoteIds, target: remoteId) {
      return true
    }
    if foundName(settings.withNames, target: advName) {
      return true
    }
    if foundKeyword(settings.withKeywords, target: advName) {
      return true
    }
    if foundMsd(settings.withMsd, msd: msd) {
      return true
    }
    if foundServiceData(settings.withServiceData, sd: serviceData) {
      return true
    }
    return false
  }

  static func foundService(_ services: [String], target: [CBUUID]?) -> Bool {
    guard let target = target, !target.isEmpty else { return false }
    let lowercased = services.map { $0.lowercased() }
    for uuid in target where lowercased.contains(uuid.uuidStr) {
      return true
    }
    return false
  }

  static func foundName(_ names: [String], target: String?) -> Bool {
    guard let target = target else { return false }
    return names.contains(target)
  }

  static func foundKeyword(_ keywords: [String], target: String?) -> Bool {
    guard let target = target else { return false }
    for keyword in keywords where target.contains(keyword) {
      return true
    }
    return false
  }

  static func foundRemoteId(_ remoteIds: [String], target: String?) -> Bool {
    guard let target = target?.lowercased() else { return false }
    for remoteId in remoteIds where remoteId.lowercased() == target {
      return true
    }
    return false
  }

  static func foundMsd(_ filters: [BmMsdFilter], msd: Data?) -> Bool {
    guard let msd = msd, msd.count >= 2 else { return false }

    // first 2 bytes are the manufacturer id (little endian)
    let bytes = [UInt8](msd.prefix(2))
    let manufId = UInt16(bytes[0]) | (UInt16(bytes[1]) << 8)

    // trim off first 2 bytes
    let trimmed = msd.subdata(in: 2..<msd.count)

    for filter in filters {
      let data = filter.data?.data ?? Data()
      var mask = filter.mask?.data ?? Data()
      if mask.isEmpty && !data.isEmpty {
        mask = Data(repeating: 0xFF, count: data.count)
      }
      if Int64(manufId) == filter.manufacturerId && findData(data, in: trimmed, mask: mask) {
        return true
      }
    }
    return false
  }

  static func foundServiceData(_ filters: [BmServiceDataFilter], sd: [CBUUID: Data]?) -> Bool {
    guard let sd = sd, !sd.isEmpty else { return false }
    for filter in filters {
      let data = filter.data.data
      var mask = filter.mask?.data ?? Data()
      if mask.isEmpty && !data.isEmpty {
        mask = Data(repeating: 0xFF, count: data.count)
      }
      let service = filter.service.lowercased()
      for (uuid, value) in sd
      where uuid.uuidStr == service && findData(data, in: value, mask: mask) {
        return true
      }
    }
    return false
  }

  /// Bitwise masked prefix comparison of `find` against `data`.
  static func findData(_ find: Data, in data: Data, mask: Data) -> Bool {
    // find & mask must be the same length; data must be long enough
    guard find.count == mask.count, data.count >= find.count else { return false }
    let f = [UInt8](find)
    let d = [UInt8](data)
    let m = [UInt8](mask)
    for i in 0..<f.count where (f[i] & m[i]) != (d[i] & m[i]) {
      return false
    }
    return true
  }
}

File: packages/bluebird_android/android/src/main/kotlin/com/lib/bluebird/Proto.kt
// Copyright 2017-2023, Charles Weinberger & Paul DeMarco.
// Copyright 2026, Nebojša Cvetković (nebkat).
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// Conversions between pigeon (Bm*) messages and Android Bluetooth objects — the codec layer.

package com.lib.bluebird

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattService
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanRecord
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.os.Build
import android.os.ParcelUuid
import java.io.ByteArrayOutputStream

/** Wire form of an error code: snake_case, shared convention with Dart (see pigeons/messages.dart). */
val BluebirdErrorCode.wire: String get() = name.lowercase()

object Proto {

    //////////////////////////////////////////
    // typed ref building

    fun attributeId(service: BluetoothGattService): BmAttributeId =
        BmAttributeId(Uuid(service.uuid), service.instanceId.toLong())

    fun attributeId(characteristic: BluetoothGattCharacteristic): BmAttributeId =
        BmAttributeId(Uuid(characteristic.uuid), characteristic.instanceId.toLong())

    /**
     * Builds the ref for [service], setting parentService when it is a
     * secondary service included by a primary one.
     */
    fun serviceRef(gatt: BluetoothGatt, service: BluetoothGattService): BmServiceRef {
        val primary = getPrimaryService(gatt, service)
        return BmServiceRef(attributeId(service), primary?.let { attributeId(it) })
    }

    fun characteristicRef(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic): BmCharacteristicRef =
        BmCharacteristicRef(serviceRef(gatt, characteristic.service), attributeId(characteristic))

    /**
     * If [service] is a secondary service, returns the primary service that
     * includes it; otherwise null.
     */
    fun getPrimaryService(gatt: BluetoothGatt, service: BluetoothGattService): BluetoothGattService? {
        // is this *already* a primary service?
        if (service.type == BluetoothGattService.SERVICE_TYPE_PRIMARY) {
            return null
        }
        // Otherwise, iterate included services until we find the primary service
        for (primary in gatt.services) {
            for (secondary in primary.includedServices) {
                if (secondary.uuid == service.uuid) {
                    return primary
                }
            }
        }
        return null
    }

    //////////////////////////////////////////
    // typed ref resolution

    fun resolveService(gatt: BluetoothGatt, id: BmAttributeId): BluetoothGattService? {
        val target = AttributeId(id)
        return gatt.services.firstOrNull { AttributeId(it) == target }
    }

    fun resolveCharacteristic(gatt: BluetoothGatt, ref: BmCharacteristicRef): BluetoothGattCharacteristic? {
        val service = resolveService(gatt, ref.service.service) ?: return null
        val target = AttributeId(ref.characteristic)
        return service.characteristics.firstOrNull { AttributeId(it) == target }
    }

    fun resolveDescriptor(gatt: BluetoothGatt, ref: BmDescriptorRef): BluetoothGattDescriptor? {
        val characteristic = resolveCharacteristic(gatt, ref.characteristic) ?: return null
        // descriptors are uuid-unique within a characteristic (instance is always 0)
        return characteristic.descriptors.firstOrNull { Uuid(it.uuid) == ref.id.uuid }
    }

    //////////////////////////////////////////
    // op keys

    //////////////////////////////////////////
    // message builders

    fun bmBluetoothDevice(device: BluetoothDevice): BmBluetoothDevice =
        BmBluetoothDevice(device.address, device.name)

    fun bmBluetoothService(gatt: BluetoothGatt, service: BluetoothGattService): BmBluetoothService {
        val characteristics = service.characteristics.map { bmBluetoothCharacteristic(it) }
        val includedServices = service.includedServices.map { included ->
            if (included.type == BluetoothGattService.SERVICE_TYPE_SECONDARY) {
                BmServiceRef(attributeId(included), attributeId(service))
            } else {
                BmServiceRef(attributeId(included), null)
            }
        }
        return BmBluetoothService(
            id = attributeId(service),
            isPrimary = service.type == BluetoothGattService.SERVICE_TYPE_PRIMARY,
            characteristics = characteristics,
            includedServices = includedServices,
        )
    }

    fun bmBluetoothCharacteristic(characteristic: BluetoothGattCharacteristic): BmBluetoothCharacteristic =
        BmBluetoothCharacteristic(
            id = attributeId(characteristic),
            descriptors = characteristic.descriptors.map { BmBluetoothDescriptor(BmAttributeId(Uuid(it.uuid), 0L)) },
            properties = bmCharacteristicProperties(characteristic.properties),
        )

    fun bmCharacteristicProperties(properties: Int): BmCharacteristicProperties =
        BmCharacteristicProperties(
            broadcast = (properties and 1) != 0,
            read = (properties and 2) != 0,
            writeWithoutResponse = (properties and 4) != 0,
            write = (properties and 8) != 0,
            notify = (properties and 16) != 0,
            indicate = (properties and 32) != 0,
            authenticatedSignedWrites = (properties and 64) != 0,
            extendedProperties = (properties and 128) != 0,
            notifyEncryptionRequired = (properties and 256) != 0,
            indicateEncryptionRequired = (properties and 512) != 0,
        )

    fun bmScanAdvertisement(device: BluetoothDevice, result: ScanResult): BmScanAdvertisement {
        val min = Int.MIN_VALUE
        val adv: ScanRecord? = result.scanRecord

        val connectable: Boolean = if (Build.VERSION.SDK_INT >= 26) { // Android 8.0, August 2017
            result.isConnectable
        } else {
            // Prior to Android 8.0, it is not possible to get if connectable.
            true
        }

        val advName = adv?.deviceName
        val txPower = adv?.txPowerLevel ?: min
        val appearance = adv?.let { getAppearanceFromScanRecord(it) } ?: 0
        val rawMsd = adv?.let { getManufacturerSpecificData(it) }

        // Manufacturer Specific Data
        val manufacturerData = HashMap<Long, ByteArray>()
        if (rawMsd != null && rawMsd.size >= 2) {
            // manufacturer ID uses little-endian order.
            val manufacturerId = (rawMsd[0].toInt() and 0xFF) or ((rawMsd[1].toInt() and 0xFF) shl 8)
            manufacturerData[manufacturerId.toLong()] = rawMsd.copyOfRange(2, rawMsd.size)
        }

        // Service Data
        val serviceData = HashMap<String, ByteArray>()
        adv?.serviceData?.forEach { (key, value) ->
            serviceData[Uuid(key.uuid).str] = value
        }

        // Service UUIDs
        val serviceUuids = adv?.serviceUuids?.map { Uuid(it.uuid).str } ?: emptyList()

        val platformName = try {
            device.name
        } catch (e: SecurityException) {
            null
        }

        return BmScanAdvertisement(
            address = device.address,
            platformName = platformName,
            advName = advName,
            connectable = connectable,
            txPowerLevel = if (txPower != min) txPower.toLong() else null,
            appearance = if (appearance != 0) appearance.toLong() else null,
            manufacturerData = manufacturerData,
            serviceData = serviceData,
            serviceUuids = serviceUuids,
            rssi = result.rssi.toLong(),
        )
    }

    fun getAppearanceFromScanRecord(adv: ScanRecord): Int {
        if (Build.VERSION.SDK_INT >= 33) { // Android 13 (August 2022)
            val map = adv.advertisingDataMap
            val bytes = map[ScanRecord.DATA_TYPE_APPEARANCE]
            if (bytes != null && bytes.size == 2) {
                val loByte = bytes[0].toInt() and 0xFF
                val hiByte = bytes[1].toInt() and 0xFF
                return hiByte * 256 + loByte
            }
            return 0
        }

        // For API Level 21+
        val bytes = adv.bytes
        var n = 0
        while (n < bytes.size) {
            val fieldLen = bytes[n].toInt() and 0xFF

            // no more or malformed data
            if (fieldLen <= 0) break

            // Ensuring we don't go past the bytes array
            if (n + fieldLen >= bytes.size) break

            val dataType = bytes[n + 1].toInt() and 0xFF

            // no more data
            if (dataType == 0) break

            // appearance type byte
            if (dataType == 0x19 && fieldLen == 3) {
                val high = (bytes[n + 3].toInt() and 0xFF) shl 8
                val low = bytes[n + 2].toInt() and 0xFF
                return high or low
            }

            n += fieldLen + 1
        }
        return 0
    }

    // The original android implementation has a bug - it does not
    // concatenate multiple MSD in the same advertisement.
    fun getManufacturerSpecificData(adv: ScanRecord): ByteArray {
        val bytes = adv.bytes
        val output = ByteArrayOutputStream()
        var n = 0
        while (n < bytes.size) {
            // layout:
            // n[0] = fieldlen
            // n[1] = datatype (MSD)
            // n[2] = manufacturerId (low)
            // n[3] = manufacturerId (high)
            // n[4] = data...
            val fieldLen = bytes[n].toInt() and 0xFF

            // no more or malformed data
            if (fieldLen <= 0) break

            // Ensuring we don't go past the bytes array
            if (n + fieldLen >= bytes.size) break

            val dataType = bytes[n + 1].toInt() and 0xFF

            // Check for Manufacturer Specific Data type (0xFF)
            // and that the field is large enough (at least 2 bytes: type + at least 1 data byte)
            if (dataType == 0xFF && fieldLen >= 2) {
                output.write(bytes, n + 2, fieldLen - 1)
            }

            n += fieldLen + 1
        }
        return output.toByteArray()
    }

    //////////////////////////////////////////
    // scan settings

    /** Converts pigeon scan settings into the native filter list. */
    fun scanFilters(settings: BmScanSettings): List<ScanFilter> = buildList {
        // services
        for (service in settings.withServices) {
            val uuid = ParcelUuid(Uuid.parse(service).value)
            add(ScanFilter.Builder().setServiceUuid(uuid).build())
        }

        // remoteIds
        for (address in settings.withRemoteIds) {
            add(ScanFilter.Builder().setDeviceAddress(address).build())
        }

        // names
        for (name in settings.withNames) {
            add(ScanFilter.Builder().setDeviceName(name).build())
        }

        // keywords
        if (Build.VERSION.SDK_INT >= 33 && settings.withKeywords.isNotEmpty()) { // Android 13 (August 2022)
            // device must advertise a name
            add(ScanFilter.Builder()
                .setAdvertisingDataType(ScanRecord.DATA_TYPE_LOCAL_NAME_SHORT).build())
            add(ScanFilter.Builder()
                .setAdvertisingDataType(ScanRecord.DATA_TYPE_LOCAL_NAME_COMPLETE).build())
        }

        // msd
        for (msd in settings.withMsd) {
            val mask = msd.mask
            add(if (mask == null || mask.isEmpty()) {
                ScanFilter.Builder().setManufacturerData(msd.manufacturerId.toInt(), msd.data).build()
            } else {
                ScanFilter.Builder().setManufacturerData(msd.manufacturerId.toInt(), msd.data, mask).build()
            })
        }

        // service data
        for (sd in settings.withServiceData) {
            val uuid = ParcelUuid(Uuid.parse(sd.service).value)
            val mask = sd.mask
            add(if (mask == null || mask.isEmpty()) {
                ScanFilter.Builder().setServiceData(uuid, sd.data).build()
            } else {
                ScanFilter.Builder().setServiceData(uuid, sd.data, mask).build()
            })
        }
    }

    /** Converts pigeon scan settings into native scan settings. */
    fun scanSettings(settings: BmScanSettings): ScanSettings {
        val builder = ScanSettings.Builder()
        builder.setScanMode(settings.androidScanMode.toInt())
        if (Build.VERSION.SDK_INT >= 26) { // Android 8.0 (August 2017)
            builder.setPhy(ScanSettings.PHY_LE_ALL_SUPPORTED)
            builder.setLegacy(settings.androidLegacy)
        }
        return builder.build()
    }

    //////////////////////////////////////////
    // enum mapping

    fun bmAdapterStateEnum(state: Int): BluetoothAdapterState = when (state) {
        BluetoothAdapter.STATE_OFF -> BluetoothAdapterState.OFF
        BluetoothAdapter.STATE_ON -> BluetoothAdapterState.ON
        BluetoothAdapter.STATE_TURNING_OFF -> BluetoothAdapterState.TURNING_OFF
        BluetoothAdapter.STATE_TURNING_ON -> BluetoothAdapterState.TURNING_ON
        else -> BluetoothAdapterState.UNKNOWN
    }

    fun bmBondStateEnum(state: Int): BluetoothBondState = when (state) {
        BluetoothDevice.BOND_NONE -> BluetoothBondState.NONE
        BluetoothDevice.BOND_BONDING -> BluetoothBondState.BONDING
        BluetoothDevice.BOND_BONDED -> BluetoothBondState.BONDED
        else -> BluetoothBondState.NONE
    }

    fun bmConnectionPriorityParse(priority: ConnectionPriority): Int = when (priority) {
        ConnectionPriority.BALANCED -> BluetoothGatt.CONNECTION_PRIORITY_BALANCED
        ConnectionPriority.HIGH -> BluetoothGatt.CONNECTION_PRIORITY_HIGH
        ConnectionPriority.LOW_POWER -> BluetoothGatt.CONNECTION_PRIORITY_LOW_POWER
    }

    fun bytesToHex(bytes: ByteArray?): String {
        if (bytes == null) return ""
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            sb.append(Character.forDigit((b.toInt() shr 4) and 0xF, 16))
            sb.append(Character.forDigit(b.toInt() and 0xF, 16))
        }
        return sb.toString()
    }
}
